Don't overwrite existing files when DOWNLOAD_ALWAYS_ASK is false

timborden opened the pull request:
Fixes #7764

Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).

Test Plan:
- Tested that the file was downloaded to the location specified in the settings
- Tested that a second downloaded file would not replace an existing file of the same name

Reviewer Checklist:

Tests


- [x] Adequate test coverage exists to prevent regressions
- [x] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [x] New files have MPL2 license header

bridiver commented on app/filtering.js:
```diff
@@ -520,14 +521,29 @@ function registerForDownloadListener (session) {
 
     const defaultPath = path.join(getSetting(settings.DOWNLOAD_DEFAULT_PATH) || getSetting(settings.DEFAULT_DOWNLOAD_SAVE_PATH) || app.getPath('downloads'), itemFilename)
-    const savePath = ((process.env.SPECTRON || (!getSetting(settings.DOWNLOAD_ALWAYS_ASK) && !item.promptForSaveLocation())) ? defaultPath : dialog.showSaveDialog(win, { defaultPath }))
+    let savePath = ((process.env.SPECTRON || (!getSetting(settings.DOWNLOAD_ALWAYS_ASK) && !item.promptForSaveLocation())) ? defaultPath : dialog.showSaveDialog(win, { defaultPath }))
```
I think there should be a single `if` statement here that wraps all of this, but can you switch `!getSetting(settings.DOWNLOAD_ALWAYS_ASK) && !item.promptForSaveLocation()` to `getSetting(settings.DOWNLOAD_ALWAYS_ASK)  || item.promptForSaveLocation()`? The former is confusing

bridiver commented on app/filtering.js:
```diff
@@ -520,14 +521,29 @@ function registerForDownloadListener (session) {
       return
     }
 
+    if (!getSetting(settings.DOWNLOAD_ALWAYS_ASK) && !item.promptForSaveLocation()) {
```
please merge into `else` for `if` above

bridiver commented on app/filtering.js:
```diff
@@ -520,14 +521,29 @@ function registerForDownloadListener (session) {
     }
 
+    if (!getSetting(settings.DOWNLOAD_ALWAYS_ASK) && !item.promptForSaveLocation()) {
+      let willNotOverwrite = false
```
double negative checks are also hard to understand `!willNotOverwrite` -> `willOverwrite`

bridiver commented on app/filtering.js:
```diff
@@ -520,14 +521,29 @@ function registerForDownloadListener (session) {
+      let matchedFilenames = 0
+      let querySavePath
+      while (!willNotOverwrite) {
+        querySavePath = (matchedFilenames ? savePath.substring(0, savePath.lastIndexOf('.') >= 0 ? savePath.lastIndexOf('.') : savePath.length) + ` (${matchedFilenames})` + savePath.substring(savePath.lastIndexOf('.') >= 0 ? savePath.lastIndexOf('.') : savePath.length) : savePath)
```
I think this will break things like `.tar.gz`. Maybe get the first index?

timborden commented:
Thanks for the feedback @bridiver 

I've refactored the code based on your suggestions.

I also dug into the extension name edge cases (like `tar.gz`). As a result I switched from the `substring()` approach to using `path.extname()`. Sadly, it didn't solve the issue, but the code is shorter and more readable.

I did some searching for possible solutions. One thought was to use the MIME:
```
> var mime = require('mime-types')
undefined
> mime.lookup('package.json')
'application/json'
> mime.extension(mime.lookup('package.json'))
'json'
```
....sadly, it didn't work for `tar.gz` files
```
> mime.lookup('v0.17.13dev.tar.gz')
false
```

It seems that the most common approach is to split on the last "."...even the OS X save prompt defaults to using the last ".":
<img width="425" alt="screen shot 2017-07-06 at 11 01 42" src="https://user-images.githubusercontent.com/422542/27906651-5ea1e54c-623c-11e7-948d-7c11df075e92.png">

Short of rolling with a whitelist of file extensions, which feels like overkill....it looks like this may be the best option.

timborden commented:
Just added tests....FYI

luixxiul commented:
@timborden would you please rebase on the latest master? thanks!

timborden commented:
Hey @luixxiul ...sorry for the relay (just changed my GitHub settings so I don't miss these notifications)

I just rebased....FYI

luixxiul commented:
@timborden thanks! I'd like to see you keep up the nice work :-)

bsclifton commented:
@bridiver did you want to re-review before this gets merged?

bsclifton commented:
Merged 😄 👌

timborden commented:
Awesome....thanks @bsclifton, @bridiver and @luixxiul
